pkg/controllers/dynakube/activegate/internal/tls/reconciler.go

```diff
@@ -0,0 +1,171 @@
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context) error {
+	if r.dk.ActiveGate().IsEnabled() && r.dk.ActiveGate().IsAutomaticTlsSecretEnabled() {
+		return r.reconcileSelfSignedTLSSecret(ctx)
+	}
+
+	if meta.FindStatusCondition(*r.dk.Conditions(), conditionType) == nil {
+		return nil
+	}
+	defer meta.RemoveStatusCondition(r.dk.Conditions(), conditionType)
+
+	return r.deleteSelfSignedTLSSecret(ctx)
+}
+
+func (r *Reconciler) reconcileSelfSignedTLSSecret(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	_, err := query.Get(ctx, types.NamespacedName{
+		Name:      r.dk.ActiveGate().GetTlsSecretName(),
+		Namespace: r.dk.Namespace,
+	})
+
+	if err != nil && k8serrors.IsNotFound(err) {
+		return r.createSelfSignedTLSSecret(ctx)
+	}
+
+	if err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) deleteSelfSignedTLSSecret(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	return query.Delete(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.dk.ActiveGate().GetTlsSecretName(),
+			Namespace: r.dk.Namespace,
+		},
+	})
+}
+
+func (r *Reconciler) createSelfSignedTLSSecret(ctx context.Context) error {
+	cert, err := certificates.New(r.timeProvider)
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	cert.Cert.DNSNames = getCertificateAltNames(r.dk.Name)
+	cert.Cert.KeyUsage = x509.KeyUsageKeyEncipherment | x509.KeyUsageDataEncipherment
+	cert.Cert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+	cert.Cert.Subject.CommonName = r.dk.Name + activeGateSelfSignedTLSCommonNameSuffix
+	cert.Cert.IPAddresses = getCertificateAltIPs(r.dk.Status.ActiveGate.ServiceIPs)
+
+	err = cert.SelfSign()
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	pemCert, pemPk, err := cert.ToPEM()
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	coreLabels := k8slabels.NewCoreLabels(r.dk.Name, k8slabels.ActiveGateComponentLabel)
+	secretData := map[string][]byte{
+		consts.TLSCrtDataName: pemCert,
+		consts.TLSKeyDataName: pemPk,
+		tlsCrtDataName:        pemCert,
+	}
+
+	secret, err := k8ssecret.Build(r.dk, r.dk.ActiveGate().GetTlsSecretName(), secretData, k8ssecret.SetLabels(coreLabels.BuildLabels()))
+	if err != nil {
+		conditions.SetSecretGenFailed(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	secret.Type = corev1.SecretTypeOpaque
+
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	err = query.Create(ctx, secret)
+	if err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), conditionType, err)
+
+		return err
+	}
+
+	conditions.SetSecretCreated(r.dk.Conditions(), conditionType, secret.Name)
+
+	return nil
+}
+
+func getCertificateAltNames(dkName string) []string {
```

Comment:
we have a similar function in extensions, maybe we can move it to util or so, https://github.com/dynatrace/dynatrace-operator/blob/c827bd288b91d9ef79f1a60fac19e34f39ed56e9/pkg/controllers/dynakube/extension/tls/reconciler.go#L142
not a blocker just an idea.

Reply:
moved to `certificates` package